File: app/src/main/java/com/tunjid/rcswitchcontrol/adapters/NSDAdapter.java
package com.tunjid.rcswitchcontrol.adapters;

import android.net.nsd.NsdServiceInfo;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.tunjid.androidbootstrap.view.recyclerview.InteractiveAdapter;
import com.tunjid.androidbootstrap.view.recyclerview.InteractiveViewHolder;
import com.tunjid.rcswitchcontrol.R;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

/**
 * Adapter for showing open NSD services
 * <p>
 * Created by tj.dahunsi on 2/4/17.
 */

public class NSDAdapter extends InteractiveAdapter<NSDAdapter.NSDViewHolder,
        NSDAdapter.ServiceClickedListener> {

    private List<NsdServiceInfo> infoList;

    public NSDAdapter(NSDAdapter.ServiceClickedListener listener, List<NsdServiceInfo> list) {
        super(listener);
        this.infoList = list;
    }

    @Override @NonNull
    public NSDViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.viewholder_nsd_list, parent, false);
        return new NSDViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NSDViewHolder holder, int position) {
        holder.bind(infoList.get(position), adapterListener);
    }

    @Override
    public int getItemCount() {
        return infoList.size();
    }

    public interface ServiceClickedListener extends InteractiveAdapter.AdapterListener {
        void onServiceClicked(NsdServiceInfo serviceInfo);

        boolean isSelf(NsdServiceInfo serviceInfo);
    }

    static class NSDViewHolder extends InteractiveViewHolder<ServiceClickedListener>
            implements View.OnClickListener {

        NsdServiceInfo serviceInfo;
        TextView textView;

        NSDViewHolder(View itemView) {
            super(itemView);
            textView = (TextView) itemView;
            itemView.setOnClickListener(this);
        }

        void bind(NsdServiceInfo info, ServiceClickedListener listener) {
            serviceInfo = info;
            adapterListener = listener;

            StringBuilder stringBuilder = new StringBuilder();
            stringBuilder.append(info.getServiceName()).append("\n")
                    .append(info.getHost().getHostAddress());

            boolean isSelf = adapterListener.isSelf(info);

            if (isSelf) stringBuilder.append(" (SELF)");

            int color = ContextCompat.getColor(itemView.getContext(), isSelf
                    ? R.color.dark_grey
                    : R.color.colorPrimary);

            textView.setTextColor(color);
            textView.setText(stringBuilder.toString());
        }

        @Override
        public void onClick(View v) {
            adapterListener.onServiceClicked(serviceInfo);
        }
    }
}

File: app/src/main/java/com/tunjid/rcswitchcontrol/adapters/NSDAdapter.kt
package com.tunjid.rcswitchcontrol.adapters

import android.net.nsd.NsdServiceInfo
import android.text.SpannableStringBuilder
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.text.scale
import com.tunjid.androidbootstrap.recyclerview.InteractiveAdapter
import com.tunjid.androidbootstrap.recyclerview.InteractiveViewHolder
import com.tunjid.rcswitchcontrol.R

/**
 * Adapter for showing open NSD services
 *
 *
 * Created by tj.dahunsi on 2/4/17.
 */

class NSDAdapter(
        listener: ServiceClickedListener,
        private val infoList: List<NsdServiceInfo>
) : InteractiveAdapter<NSDAdapter.NSDViewHolder, NSDAdapter.ServiceClickedListener>(listener) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NSDViewHolder {
        return NSDViewHolder(getItemView(R.layout.viewholder_nsd_list, parent), adapterListener)
    }

    override fun onBindViewHolder(holder: NSDViewHolder, position: Int) =
            holder.bind(infoList[position], adapterListener)

    override fun getItemCount(): Int = infoList.size

    interface ServiceClickedListener : AdapterListener {
        fun onServiceClicked(serviceInfo: NsdServiceInfo)

        fun isSelf(serviceInfo: NsdServiceInfo): Boolean
    }

    class NSDViewHolder internal constructor(
            itemView: View,
            listener: ServiceClickedListener
    ) : InteractiveViewHolder<ServiceClickedListener>(itemView, listener) {

        private lateinit var serviceInfo: NsdServiceInfo
        private val textView: TextView = itemView as TextView

        init {
            itemView.setOnClickListener { adapterListener.onServiceClicked(serviceInfo) }
        }

        internal fun bind(info: NsdServiceInfo, listener: ServiceClickedListener) {
            serviceInfo = info
            adapterListener = listener

            textView.text = SpannableStringBuilder()
                    .append(info.serviceName)
                    .apply { if (adapterListener.isSelf(info)) append(" (SELF)") }
                    .append("\n")
                    .scale(0.8F) { append(info.host.hostAddress) }
        }

    }
}
